File: 04-passportProcessing/partTwo.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	inputBytes, _ := ioutil.ReadFile("input.txt")
	inputSlice := strings.Split(strings.TrimSpace(string(inputBytes)), "\n\n")

	for i, x := range inputSlice {
		inputSlice[i] = strings.ReplaceAll(x, "\n", " ")
	}

	var passports []Passport
	for _, x := range inputSlice {
		passports = append(passports, NewPassport(x))
	}

	var validPassports int
	for _, passport := range passports {
		if passport.Validate() {
			validPassports += 1
		}
	}

	fmt.Printf("There are %d valid passports.", validPassports)

}

type Passport struct {
	byr string // Birth year
	iyr string // Issue year
	eyr string // Expiration year
	hgt string // Height
	hcl string // Hair colour
	ecl string // Eye colour
	pid string // Passport ID
	cid string // Country ID
}

func (p Passport) Validate() bool {
	// byr (Birth Year) - four digits; at least 1920 and at most 2002.
	if p.byr == "" {
		return false
	}
	{
		i, _ := strconv.Atoi(p.byr)
		if !(1920 <= i && i <= 2002) {
			return false
		}
	}

	// iyr (Issue Year) - four digits; at least 2010 and at most 2020.
	if p.iyr == "" {
		return false
	}
	{
		i, _ := strconv.Atoi(p.iyr)
		if !(2010 <= i && i <= 2020) {
			return false
		}
	}

	// eyr (Expiration Year) - four digits; at least 2020 and at most 2030.
	if p.eyr == "" {
		return false
	}
	{
		i, _ := strconv.Atoi(p.eyr)
		if !(2020 <= i && i <= 2030) {
			return false
		}
	}

	// hgt (Height) - a number followed by either cm or in:
	//     If cm, the number must be at least 150 and at most 193.
	//     If in, the number must be at least 59 and at most 76.
	if p.hgt == "" {
		return false
	}
	if strings.Contains(p.hgt, "cm") {
		i, _ := strconv.Atoi(strings.Trim(p.hgt, "cm"))
		if !(150 <= i && i <= 193) {
			return false
		}
	} else if strings.Contains(p.hgt, "in") {
		i, _ := strconv.Atoi(strings.Trim(p.hgt, "in"))
		if !(59 <= i && i <= 76) {
			return false
		}
	} else {
		return false
	}

	// hcl (Hair Color) - a // followed by exactly six characters 0-9 or a-f.
	if p.hcl == "" {
		return false
	}
	if p.hcl[0] == []byte("#")[0] {
		if m, _ := regexp.MatchString(`#[0-9a-f]{6}`, p.hcl); !m {
			return false
		}
	} else {
		return false
	}

	// ecl (Eye Color) - exactly one of: amb blu brn gry grn hzl oth.
	if p.ecl == "" {
		return false
	}
	{
		var contained bool
		for _, v := range []string{"amb", "blu", "brn", "gry", "grn", "hzl", "oth"} {
			if p.ecl == v {
				contained = true
				break
			}
		}
		if !contained {
			return false
		}
	}

	// pid (Passport ID) - a nine-digit number, including leading zeroes.
	if len(p.pid) == 9 {
		_, err := strconv.Atoi(p.pid)
		if err != nil {
			return false
		}
	} else {
		return false
	}

	return true
}

func NewPassport(instr string) Passport {
	np := Passport{}
	np.byr = extractField("byr", instr)
	np.iyr = extractField("iyr", instr)
	np.eyr = extractField("eyr", instr)
	np.hgt = extractField("hgt", instr)
	np.hcl = extractField("hcl", instr)
	np.ecl = extractField("ecl", instr)
	np.pid = extractField("pid", instr)
	np.cid = extractField("cid", instr)
	return np
}

func extractField(field, instr string) string {
	fieldRegex := regexp.MustCompile(field + `:([^ ]+)`)
	matches := fieldRegex.FindAllStringSubmatch(instr, -1)
	if len(matches) == 0 {
		return ""
	}
	return matches[0][1]
}

File: 04-passportProcessing/partTwo.py
import re

input_list = [x.replace("\n", " ") for x in open("input.txt").read().strip().split("\n\n")]

class Passport: 
    byr: str  # Birth year
    iyr: str  # Issue year
    eyr: str  # Expiration year
    hgt: str  # Height
    hcl: str  # Hair colour
    ecl: str  # Eye colour
    pid: str  # Passport ID
    cid: str  # Country ID

    def __init__(self, instr:str) -> None:
        self.byr = self._extract_field("byr", instr)
        self.iyr = self._extract_field("iyr", instr)
        self.eyr = self._extract_field("eyr", instr)
        self.hgt = self._extract_field("hgt", instr)
        self.hcl = self._extract_field("hcl", instr)
        self.ecl = self._extract_field("ecl", instr)
        self.pid = self._extract_field("pid", instr)
        self.cid = self._extract_field("cid", instr)

    def _extract_field(self, field:str, instr:str) -> str:
        matches = re.search(field + r":([^ ]+)", instr)
        if matches is None:
            return ""
        
        return matches.group(1)

    def validate(self) -> bool:
        # byr (Birth Year) - four digits; at least 1920 and at most 2002.
        if self.byr == "":
            return False
        if not (1920 <= int(self.byr) <= 2002):
            return False

        # iyr (Issue Year) - four digits; at least 2010 and at most 2020.
        if self.iyr == "":
            return False
        if not (2010 <= int(self.iyr) <= 2020):
            return False
        
        # eyr (Expiration Year) - four digits; at least 2020 and at most 2030.
        if self.eyr == "":
            return False
        if not (2020 <= int(self.eyr) <= 2030):
            return False

        # hgt (Height) - a number followed by either cm or in:
        #     If cm, the number must be at least 150 and at most 193.
        #     If in, the number must be at least 59 and at most 76.
        if self.hgt == "":
            return False
        if "cm" in self.hgt:
            if not (150 <= int(self.hgt.strip("cm")) <= 193):
                return False
        elif "in" in self.hgt:
            if not (59 <= int(self.hgt.strip("in")) <= 76):
                return False
        else:
            return False

        # hcl (Hair Color) - a # followed by exactly six characters 0-9 or a-f.
        if self.hcl == "":
            return False
        if self.hcl[0] == "#":
            if re.match(r"#[0-9a-f]{6}", self.hcl) is None:
                return False
        else:
            return False

        # ecl (Eye Color) - exactly one of: amb blu brn gry grn hzl oth.
        if self.ecl == "":
            return False
        if self.ecl not in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]:
            return False

        # pid (Passport ID) - a nine-digit number, including leading zeroes.
        if len(self.pid) == 9:
            try:
                int(self.pid)
            except ValueError:
                return False
        else:
            return False

        return True

passports = [Passport(x) for x in input_list]

valid_passports = 0
for passport in passports:
    if passport.validate():
        valid_passports += 1

print(f"There are {valid_passports} valid passports.")
